master/internal/api_experiment.go

```diff
@@ -99,44 +80,100 @@ func (a *apiServer) GetExperiment(
 
 func (a *apiServer) GetExperiments(
 	_ context.Context, req *apiv1.GetExperimentsRequest) (*apiv1.GetExperimentsResponse, error) {
+	// Construct the experiment filtering expression.
+	var allStates []string
+	for _, state := range req.States {
+		allStates = append(allStates, strings.TrimPrefix(state.String(), "STATE_"))
+	}
+	stateFilterExpr := strings.Join(allStates, ",")
+	userFilterExpr := strings.Join(req.Users, ",")
+	labelFilterExpr := strings.Join(req.Labels, ",")
+	archivedExpr := ""
+	if req.Archived != nil {
+		archivedExpr = strconv.FormatBool(req.Archived.Value)
+	}
+
+	// Construct the ordering expression.
+	orderColMap := map[apiv1.GetExperimentsRequest_SortBy]string{
+		apiv1.GetExperimentsRequest_SORT_BY_UNSPECIFIED: "id",
+		apiv1.GetExperimentsRequest_SORT_BY_ID:          "id",
+		apiv1.GetExperimentsRequest_SORT_BY_DESCRIPTION: "description",
+		apiv1.GetExperimentsRequest_SORT_BY_START_TIME:  "start_time",
+		apiv1.GetExperimentsRequest_SORT_BY_END_TIME:    "end_time",
+		apiv1.GetExperimentsRequest_SORT_BY_STATE:       "state",
+		apiv1.GetExperimentsRequest_SORT_BY_NUM_TRIALS:  "num_trials",
+		apiv1.GetExperimentsRequest_SORT_BY_PROGRESS:    "progress",
+		apiv1.GetExperimentsRequest_SORT_BY_USER:        "username",
+	}
+	sortByMap := map[apiv1.OrderBy]string{
+		apiv1.OrderBy_ORDER_BY_UNSPECIFIED: "ASC",
+		apiv1.OrderBy_ORDER_BY_ASC:         "ASC",
+		apiv1.OrderBy_ORDER_BY_DESC:        "DESC",
+	}
+	orderExpr := ""
+	if orderColMap[req.SortBy] != "id" {
+		orderExpr = fmt.Sprintf(
+			"%s %s, id %s",
+			orderColMap[req.SortBy], sortByMap[req.OrderBy], sortByMap[req.OrderBy],
+		)
+	} else {
+		orderExpr = fmt.Sprintf("id %s", sortByMap[req.OrderBy])
+	}
+
+	// Query the database.
 	resp := &apiv1.GetExperimentsResponse{}
-	if err := a.m.db.QueryProto("get_experiments", &resp.Experiments); err != nil {
+	type experimentsWrapper struct {
+		Count       int32  `db:"count"`
+		Experiments []byte `db:"experiments"`
+	}
+	res := experimentsWrapper{}
+	if err := a.m.db.QueryF(
+		"get_experiments",
+		[]interface{}{orderExpr},
+		&res,
+		stateFilterExpr,
+		archivedExpr,
+		userFilterExpr,
+		labelFilterExpr,
+		req.Description,
+		req.Offset,
+		req.Limit,
+	); err != nil {
 		return nil, err
 	}
-	a.filter(&resp.Experiments, func(i int) bool {
-		v := resp.Experiments[i]
-		if req.Archived != nil && req.Archived.Value != v.Archived {
-			return false
+	if len(res.Experiments) > 0 {
+		unmarshaled := make([]interface{}, 0)
```

Comment:
I feel there has to be a way to do this all in one `json.Unmarshal` call into a `[]*experimentv1.Experiment`, especially instead of unmarshaling into `[]interface{}` then marshaling the `interface{}` again and unmarshaling it into `*experimentv1.Experiment`. Why is the code this way?

Reply:
`[]*experimentv1.Experiment` cannot be unmarshaled by `protojson.Unmarshal`. We used the same logic in `QueryProto`. I couldn't find any functions in `protojson` to replace the unmarshalling logic here.

Reply:
Ah I see.. that is very unfortunate.